# Processamento dos Dados e Análise Exploratória

In [25]:
### Importação dos packages necesssários

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Leitura do ficheiro

In [26]:
data_path = r'C:\Users\catarina.miranda\PycharmProjects\FicheiroCompras_ML\data\raw\JE_Valérius_Sede_Compras_2024.xlsx'

data = pd.read_excel(data_path)
print('Ficheiro lido')

Ficheiro lido


### Análise inicial dos dados

In [27]:
# Primeiro vamos analisar o cabeçalho do ficheiro
data.head(10)

Unnamed: 0                                 Afornecedo  \
0         NaN              INFORCAVADO-INFORMATICA, LDA.   
1         NaN                                         RH   
2         NaN                                         RH   
3         NaN                                         RH   
4         NaN                                         RH   
5         NaN  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
6         NaN  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
7         NaN  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
8         NaN  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
9         NaN  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   

                                   Designação Fatura  Centro ana Data fisca  \
0  integração do sistema desenvolvido no âmbito d...         NaN 2024-11-12   
1                                          Custos RH         NaN        NaT   
2                                          Custos RH         NaN        NaT   
3                                          Custos RH         NaN        NaT   
4                                          Custos RH         NaN        NaT   
5      Gasóleo para Pesados de Mercadorias 100% Ded.         1.0 2024-01-01   
6      Gasóleo para Pesados de Mercadorias 100% Ded.         1.0 2024-01-01   
7      Gasóleo para Pesados de Mercadorias 100% Ded.         1.0 2024-01-01   
8      Gasóleo para Pesados de Mercadorias 100% Ded.         1.0 2024-01-01   
9      Gasóleo para Pesados de Mercadorias 100% Ded.         1.0 2024-01-01   

   Euros  deb  Euros  cre  saldo       Conta  \
0    85000.00         NaN    NaN         NaN   
1    18175.54         NaN    NaN         NaN   
2    22282.90         NaN    NaN         NaN   
3    37948.03         NaN    NaN         NaN   
4    61208.66         NaN    NaN         NaN   
5        0.00       17.30  17.30  62421413.0   
6        0.00       17.28  17.28  62421413.0   
7        0.00       17.28  17.28  62421413.0   
8        0.00       17.28  17.28  62421413.0   
9        0.00       17.28  17.28  62421413.0   

                                  Nome da co  ... Doc  inici Numero do   \
0                                        NaN  ...        NaN        NaN   
1                                        NaN  ...        NaN        NaN   
2                                        NaN  ...        NaN        NaN   
3                                        NaN  ...        NaN        NaN   
4                                        NaN  ...        NaN        NaN   
5  Gasóleo  Pessado Mercadorias MN Tx Normal  ...         PT  1000333.0   
6  Gasóleo  Pessado Mercadorias MN Tx Normal  ...         PT  1000333.0   
7  Gasóleo  Pessado Mercadorias MN Tx Normal  ...         PT  1000333.0   
8  Gasóleo  Pessado Mercadorias MN Tx Normal  ...         PT  1000333.0   
9  Gasóleo  Pessado Mercadorias MN Tx Normal  ...         PT  1000333.0   

    Rubrica                                         Descritivo   Doc  nome1  \
0       NaN  integração do sistema desenvolvido no âmbito d...          NaN   
1       NaN                                          Custos RH          NaN   
2       NaN                                          Custos RH          NaN   
3       NaN                                          Custos RH          NaN   
4       NaN                                          Custos RH          NaN   
5  28-04-RR  V/Nt. Cré NCC1140-23A1/13 - PETRO AGRELA COMER...  Compras OBS   
6  AD-89-DP  V/Nt. Cré NCC1140-23A1/13 - PETRO AGRELA COMER...  Compras OBS   
7  AC-76-FT  V/Nt. Cré NCC1140-23A1/13 - PETRO AGRELA COMER...  Compras OBS   
8  AO-69-DQ  V/Nt. Cré NCC1140-23A1/13 - PETRO AGRELA COMER...  Compras OBS   
9  15-54-JJ  V/Nt. Cré NCC1140-23A1/13 - PETRO AGRELA COMER...  Compras OBS   

  Elegível? Alinhamento Atividade/código CAPEX/OPEX/Volume Negócios  \
0       Sim         Não          9.1 MAC                      CAPEX   
1       Sim         Não          5.5 MAC                       OPEX   
2       Sim         Não          5

In [28]:
# Vamos começar por eliminar colunas e linhas que não nos interessem (A escolha das colunas a remover foi realizada de forma iterativa, ao mesmo tempo em que era realizada a                                                                           análise exploratória que se segue)

remover_colunas = [
    'Unnamed: 0',
    'Alinhamento',
    'Data fisca',
    'Rubrica',
    'Atividade/código',
    'CAPEX/OPEX/Volume Negócios',
    'OBSERVAÇÕES',
    'saldo',
    'Euros  deb',
    'Euros  cre',
    'Centro ana',
    'Doc  n  do',
    'Numero do ',
    'Descritivo']

remover_linhas = [0,1,2,3,4]

dataset = data.drop(columns=remover_colunas)
dataset = dataset.drop(remover_linhas)

In [29]:
dataset.head(10)

Afornecedo  \
5   PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
6   PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
7   PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
8   PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
9   PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
10                JOSE CARDOSO & FERREIRA LDA   
11       RDUZ GESTÃO GLOBAL DE RESÍDUOS, S.A.   
12                         VIAGENS ABREU, SA.   
13                         VIAGENS ABREU, SA.   
14       RDUZ GESTÃO GLOBAL DE RESÍDUOS, S.A.   

                                    Designação Fatura       Conta  \
5       Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
6       Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
7       Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
8       Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
9       Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
10                Ferramentas e utensílios - Iva ded.   6231113.0   
11  Serviço Gestão Resíduos (Aluguer/Recolha/Trata...   6268113.0   
12                                               Voos   6251514.0   
13                                               Voos   6251416.0   
14  Serviço Gestão Resíduos (Aluguer/Recolha/Trata...   6268111.0   

                                   Nome da co          Doc  nome  N  contrib  \
5   Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
6   Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
7   Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
8   Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
9   Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
10                 Ferra e Utens MN Tx Normal        V/ Factura S  505215020   
11                      Resíduos MN Tx Normal        V/ Factura S  507225740   
12                             Voos MN Isento        V/ Factura S  500297177   
13         Estadia/Alojamento MN Não Ded Base        V/ Factura S  500297177   
14                    Resíduos MN Tx Reduzida        V/ Factura S  507225740   

   Doc  inici   Doc  nome1 Elegível?  
5          PT  Compras OBS       Não  
6          PT  Compras OBS       Não  
7          PT  Compras OBS       Não  
8          PT  Compras OBS       Não  
9          PT  Compras OBS       Não  
10         PT  Compras OBS       Não  
11         PT  Compras OBS       Sim  
12         PT  Compras OBS       Não  
13         PT  Compras OBS       Não  
14         PT  Compras OBS       Sim

O dataset já não apresenta colunas/linhas que **não** sejam interessantes para a análise.

In [30]:
# Verificação de NAs (Valores Nulos / Vazios)

dataset.isna().sum()


Afornecedo           3268
Designação Fatura       0
Conta                   0
Nome da co              0
Doc  nome               0
N  contrib            110
Doc  inici              0
Doc  nome1              0
Elegível?               0
dtype: int64

A coluna "Afornecedo" possui 3268 NAs e a coluna "N contrib" 110

Abordagem definida para esses casos:

Verificámos que existe vários problemas entre a interação de **N  contrib** e **Afornecedo**. Isso porque, em muitos casos identificados, o contribuinte corresponde ao comprador, e noutros casos corresponde ao vendedor, o que faz com que a sua relação não seja lógica nem padronizada.


- **"Afornecedo"** -> As linhas com NAs serão **eliminadas** (Apesar de serem 3mil linhas, não causam impacto no target como é possível verificar em baixo)

- **"N contrib"** -> As linhas com NAs serão **eliminadas**


In [31]:
#dataset["Afornecedo"] = dataset["Afornecedo"].fillna("Valérius")

print(dataset["Elegível?"].value_counts())

dataset = dataset.dropna(subset=["N  contrib"])
dataset = dataset.dropna(subset=["Afornecedo"])

dataset.isna().sum()
print(dataset["Elegível?"].value_counts())


Elegível?
Não    17536
Sim      672
Name: count, dtype: int64
Elegível?
Não    14158
Sim      672
Name: count, dtype: int64


Já não existem valores nulos/inválidos no dataset.

In [32]:
# Vamos transformar a variável binária categórica ('Elegível?') numa variável binária numérica. Assim é mais fácil para o modelo!

dataset['Elegível?'] = dataset['Elegível?'].str.lower().str.strip()
dataset['Elegível?'] = dataset['Elegível?'].map({'sim':1, 'não':0, 'nao':0})

dataset['Elegível?'].value_counts()

Elegível?
0    14158
1      672
Name: count, dtype: int64

Também é importante eliminar ruído existente em algumas colunas:

In [33]:
dataset['Doc  inici'].value_counts() # Visto que 14k das linhas possuem PT nesta coluna, tem sentido categorizar as restantes                                         como "Outros"

dataset['Doc  inici'] = dataset['Doc  inici'].apply(lambda x: x if x == 'PT' else 'Outros')

dataset['Doc  inici'].value_counts()

Doc  inici
PT        14320
Outros      510
Name: count, dtype: int64

In [34]:
# Verificação do dataset para ver se está tudo bem
dataset.head()

Afornecedo  \
5  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
6  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
7  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
8  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
9  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   

                               Designação Fatura       Conta  \
5  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
6  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
7  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
8  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
9  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   

                                  Nome da co          Doc  nome  N  contrib  \
5  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
6  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
7  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
8  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
9  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   

  Doc  inici   Doc  nome1  Elegível?  
5         PT  Compras OBS          0  
6         PT  Compras OBS          0  
7         PT  Compras OBS          0  
8         PT  Compras OBS          0  
9         PT  Compras OBS          0

**(Opcional)** Por fim, vamos guardar o dataframe no formato csv!

In [35]:
dataset.to_csv(r'C:\Users\catarina.miranda\PycharmProjects\FicheiroCompras_ML\data\processed\Ficheiro_Compras_Processado.csv', index=False)

# Análise Exploratória - Pós Processamento

### Revisão geral

Antes de tentar encontrar alguma relação entre as variáveis vamos rever os datasets.

In [36]:
print('Dataset após o processamento:')
print(f'Linhas: {dataset.shape[0]}')
print(f'Colunas: {dataset.shape[1]}')

Dataset após o processamento:
Linhas: 14830
Colunas: 9


In [37]:
print('Dataset antes vs Dataset depois de ser processado:')

print(f'Dataset antes -> {data.shape}')
print(f'Dataset depois -> {dataset.shape}')
print(f'Atualmente o dataset tem menos {data.shape[0] - dataset.shape[0]} linhas e menos {data.shape[1] - dataset.shape[1]} colunas')

Dataset antes vs Dataset depois de ser processado:
Dataset antes -> (18213, 23)
Dataset depois -> (14830, 9)
Atualmente o dataset tem menos 3383 linhas e menos 14 colunas


In [38]:
print(f'As colunas eliminadas foram: {remover_colunas}')

As colunas eliminadas foram: ['Unnamed: 0', 'Alinhamento', 'Data fisca', 'Rubrica', 'Atividade/código', 'CAPEX/OPEX/Volume Negócios', 'OBSERVAÇÕES', 'saldo', 'Euros  deb', 'Euros  cre', 'Centro ana', 'Doc  n  do', 'Numero do ', 'Descritivo']


Vamos verificar se ainda existe diferença entre os valores apresentados em **Afornecedo** e **N  contrib**


In [39]:
dataset["Afornecedo"].value_counts()

Afornecedo
ÉRIUS - TEXTEIS, SA.                               1984
INFRAESTRUTURAS DE PORTUGAL, S.A.                  1765
ALVES BANDEIRA & CA S.A.                           1029
Fornecedor Compras a Dinheiro MN                    551
BRISA - CONCESSÃO RODOVIÁRIA, S.A.                  429
                                                   ... 
MEDIUM - SOCIEDADE DE MEDIAÇÃO IMOBILIÁRIA, LDA       1
HIPOLITO & BARROS, LDA                                1
IBEROKI - PRODUTOS QUÍMICOS, LDA                      1
Guangzhou Marigold trading co. LTD                    1
IMBOTEX S.R.L.                                        1
Name: count, Length: 492, dtype: int64

In [40]:
dataset["N  contrib"].value_counts()

N  contrib
509577547      1984
503933813      1765
500433402      1029
502790024       429
507165608       398
               ... 
503644161         1
501139184         1
01662310976       1
510852360         1
504520296         1
Name: count, Length: 574, dtype: int64

In [41]:
dataset[["N  contrib", "Afornecedo"]]

N  contrib                                 Afornecedo
5      503566454  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA
6      503566454  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA
7      503566454  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA
8      503566454  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA
9      503566454  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA
...          ...                                        ...
18208  500297177                         VIAGENS ABREU, SA.
18209  210938897         MARIA ADELINA DIAS DE M. FERNANDES
18210  513003495                 CONFEÇOES LINHA LONGA, LDA
18211   37667371                           VALERIUS MOROCCO
18212   37667371                           VALERIUS MOROCCO

[14830 rows x 2 columns]

Não há evidências de não correlação entre fornecedor e número de contribuinte.

### Influência de algumas das colunas no target
Vamos analisar algumas colunas que parecem ser das mais interessantes:

In [42]:
colunas_texto = ['Designação Fatura', 'Descritivo', 'Nome da co']

print("=== ANÁLISE DAS COLUNAS DE TEXTO ===")
for coluna in colunas_texto:
    if coluna in dataset.columns:
        print(f"\n--- {coluna} ---")
        print(f"Valores únicos: {dataset[coluna].nunique()}")
        print(f"Missing values: {dataset[coluna].isnull().sum()}")
        print("Exemplos:")
        print(dataset[coluna].dropna().head(3).tolist())

=== ANÁLISE DAS COLUNAS DE TEXTO ===

--- Designação Fatura ---
Valores únicos: 1384
Missing values: 0
Exemplos:
['Gasóleo para Pesados de Mercadorias 100% Ded.', 'Gasóleo para Pesados de Mercadorias 100% Ded.', 'Gasóleo para Pesados de Mercadorias 100% Ded.']

--- Nome da co ---
Valores únicos: 154
Missing values: 0
Exemplos:
['Gasóleo  Pessado Mercadorias MN Tx Normal', 'Gasóleo  Pessado Mercadorias MN Tx Normal', 'Gasóleo  Pessado Mercadorias MN Tx Normal']


In [43]:
nome_co_stats = dataset.groupby('Nome da co').agg({
    'Elegível?': ['sum', 'mean']
}).round(3)

nome_co_stats.sort_values(by=('Elegível?', 'mean'), ascending=False, inplace=True)

nome_co_stats

Elegível?       
                                              sum   mean
Nome da co                                              
Cons Edf. e O. Const. MN Tx Reduzida            1  1.000
Equip. Bás. Máquinas MC Tx Normal               2  1.000
Resíduos MN Tx Normal                          12  1.000
Cons Veic. Lig Mercadorias MN Tx Normal       309  0.981
Transp Mercad Normal MN Tx Isento             193  0.915
...                                           ...    ...
Ensaios/Testes MN Tx Normal                     0  0.000
Ensaios/Testes OM Isento                        0  0.000
Eq. Inf./ Software MN - Taxa Normal             0  0.000
Estadia/Alojamento MN Não Ded Base              0  0.000
Água MN Tx Reduzida                             0  0.000

[154 rows x 2 columns]

Neste caso é possível observar em que tipos de **"Nome da co"** mais influenciam o **target**, nomeadamente:
- Cons Edf. e O. Const. MN Tx Reduzida;
- Equip. Bás. Máquinas MC Tx Normal
- Resíduos MN Tx Normal
- Cons Veic. Lig Mercadorias MN Tx Normal
- Transp Mercad Normal MN Tx Isento
- Cons Veic. Lig Passageiros MN Base Não Ded
- Cons Veic. Lig Passageiros MN Não Ded Tx Normal
- Transp Mercad Normal MN Tx Normal
- Cons Veic. Pes Mercadorias MN Tx Normal

É de destacar que algumas das colunas possuem uma **elevada taxa de eligibilidade,mesmo aparecendo diversas vezes**.


Vamos fazer o mesmo, mas para a coluna **'Doc  nome1'**:

In [44]:
doc_nome_values = dataset['Doc  nome ']

doc_nome_stats = dataset.groupby('Doc  nome ').agg({
    'Elegível?': ['sum', 'mean']
}).round(3)

doc_nome_stats.sort_values(by=('Elegível?', 'mean'), ascending=False, inplace=True)

print(doc_nome_values)
doc_nome_stats

5        V/ Nota de Crédito
6        V/ Nota de Crédito
7        V/ Nota de Crédito
8        V/ Nota de Crédito
9        V/ Nota de Crédito
                ...        
18208          V/ Factura S
18209            V/ Factura
18210            V/ Factura
18211            V/ Factura
18212            V/ Factura
Name: Doc  nome , Length: 14830, dtype: object


Elegível?       
                         sum   mean
Doc  nome                          
V/ Vd. Dinheiro           51  0.073
V/ Factura S             615  0.072
V/ Factura                 6  0.001
V/ Nota de Crédito         0  0.000

Nenhuma das colunas se destaca relativamente à sua ligação com o target.

E novamente para a coluna **'Conta'**

In [45]:
conta_values = dataset['Conta'].value_counts()

conta_stats = dataset.groupby('Conta').agg({
    'Elegível?': ['sum', 'mean']
}).round(3)

conta_stats.sort_values(by=('Elegível?', 'mean'), ascending=False, inplace=True)

print(conta_values)
conta_stats

Conta
62421115.0    1503
62421113.0    1503
62511113.0    1407
312113.0       957
621213.0       715
              ... 
626306.0         1
6231112.0        1
6251815.0        1
621612.0         1
6221123.0        1
Name: count, Length: 154, dtype: int64


Elegível?       
                  sum   mean
Conta                       
6268113.0          12  1.000
62263123.0          2  1.000
6226211.0           1  1.000
62264113.0        309  0.981
6253114.0         193  0.915
...               ...    ...
626306.0            0  0.000
626307.0            0  0.000
626308.0            0  0.000
626513.0            0  0.000
626822113.0         0  0.000

[154 rows x 2 columns]

É possível perceber que algumas das entradas da coluna **'Conta'** apresentam uma **relação significativa** com a **eligibilidade do investimento**, nomeadamente:
- 6268113.0
- 62263123.0
- 6226211.0
- **62264113.0**
- **6253114.0**
- 62264215.0
- 62264216.0
- 6253113.0
- 62264313.0

As entradas da coluna "Conta" que se encontram realçadas, apresentam uma taxa de eligibilidade elevada aomesmo tempo em que aparecem de forma frequente.

In [46]:
country_values = dataset['Doc  inici'].value_counts()

country_stats = dataset.groupby('Doc  inici').agg({
    'Elegível?': ['sum', 'mean']
}).round(3)

country_stats.sort_values(by=('Elegível?', 'mean'), ascending=False, inplace=True)

print(country_values)
country_stats

Doc  inici
PT        14320
Outros      510
Name: count, dtype: int64


Elegível?       
                 sum   mean
Doc  inici                 
PT               669  0.047
Outros             3  0.006

### Verificação do tipo de dados das colunas

In [47]:
dataset.dtypes

Afornecedo            object
Designação Fatura     object
Conta                float64
Nome da co            object
Doc  nome             object
N  contrib            object
Doc  inici            object
Doc  nome1            object
Elegível?              int64
dtype: object

In [49]:
dataset.head()

Afornecedo  \
5  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
6  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
7  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
8  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   
9  PETRO AGRELA COMERCIO DE COMBUSTIVEIS LDA   

                               Designação Fatura       Conta  \
5  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
6  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
7  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
8  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   
9  Gasóleo para Pesados de Mercadorias 100% Ded.  62421413.0   

                                  Nome da co          Doc  nome  N  contrib  \
5  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
6  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
7  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
8  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   
9  Gasóleo  Pessado Mercadorias MN Tx Normal  V/ Nota de Crédito  503566454   

  Doc  inici   Doc  nome1  Elegível?  
5         PT  Compras OBS          0  
6         PT  Compras OBS          0  
7         PT  Compras OBS          0  
8         PT  Compras OBS          0  
9         PT  Compras OBS          0